contrib/design-docs/config-file-parsing.md

```diff
@@ -0,0 +1,342 @@
+other files.
+
+##### storage.conf
+
+Deprecate `rootless_storage_path` option. With the `rootless/rootful` config location and admin could just use
+`graphroot` in the location in the rootless file location. As such there is no need to special case these fields
+in the parser.
+
+String arrays in the config will need to get switched to the attributedstring type, as described under Appending arrays.
+
+##### registries.conf
+
+Remove V1 config layout to simplify the parsing logic. If we do major config changes we might as well take the
+chance and remove this old format.
+Currently the V1 format is already rejected for drop-in files so this just effects the main config file.
+
+Additionally there might be a few challenges here, c/image uses the SystemContext struct which allows
+users to set `RootForImplicitAbsolutePaths`, `SystemRegistriesConfPath`, `SystemRegistriesConfDirPath`.
+We must continue to support them as they are used by various tools.
+For `RootForImplicitAbsolutePaths` we update it to check bot the `/usr` and `/etc` locations.
+When `SystemRegistriesConfPath` or `SystemRegistriesConfDirPath` are used don't do the normal parsing
+and just read the file/directory specified there.
+
+As described under the Environment Variables section the handling for `CONTAINERS_REGISTRIES_CONF` is moved
+out of Podman, and common/libimage into the actual `configfile` package. As such all users of c/image will
+be able to use this without having each caller specify their own env.
+As part of this I propose removing support for the old `REGISTRIES_CONFIG_PATH` env which was never documented
+and replaced by `CONTAINERS_REGISTRIES_CONF` in commit c9ef2607104a0b17e5146b3ee01852edb7d3d688 (over 4 years ago).
+Currently there is an issue because Buildah doesn't support it [4].
+
+String arrays in the config will need to get switched to the attributedstring type, as described under Appending arrays.
+
+#### Maybe in scope?
+
+##### registries.d
+
+Note registries.d is confusingly a completely different config file format from registries.conf.d.
+
+It is unclear to me what we should do with this. Do we want to adopt the drop in logic as explained above?
+Right now it only uses `$HOME/.config/containers/registries.d` or `/etc/containers/registries.d`
+
+I think for consistency it would be best if it uses all drop in paths like shown above.
+At least we should add `/usr/share/containers` lookup locations.
+
+Additionally there seems to exit code duplication in `podman/pkg/trust`. We should find a way to
+not duplicate this logic at all.
+
+##### certs.d
+
+Same point as for registries.d. Should we make it also use all paths?
+There is also an open issue for proper XDG_CONFIG_HOME support [2].
+
+##### policy.json
+
+This is a json file so I don't think the normal drop-in logic would be particular useful here. Though
+I think it should read at least also `/usr/share/containers/policy.json`.
```

Comment:
The worry here is “If you make a typo when installing `/etc/containers/policy.json`, your configuration is going to be ignored”.

I could see an argument that this is no different from any other config file… yet, right now, on a typo, the `policy.json` mechanism “fails closed” and forces the user to fix that before being able to pull images. (Admittedly there are other typos where the “fails closed” property can’t be upheld, e.g. a typo when installing an _updated_ policy is undetectable by c/image.)

Reply:
Well the obvious argument is you have the exact same problem today.
If I typo the path for my custom homedir config it falls back to the /etc one which I think basically on all distro which ship is is a accept anything? So really adding one more location is not going to change that?

Users/Admins are responsible for configuring the systems we cannot protect for so things, `podman image trust show` outputs the policy so admin should verify that and I guess log-level debug etc...
If you have an important policy it is always the end user would should actually tests that the policy gets enforced by us.